code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -73,36 +113,173 @@ public void computeInstanceActivateReply(Event event) {
         }
     }
 
-    protected boolean handleUnreportedKnownInstances(long agentId, Map<String, String> knownInstances,
-            ReportedInstances reportedInstances, boolean checkOnly) {
-
-        if (knownInstances != null) {
-            for (Map.Entry<String, String> knownInstance : knownInstances.entrySet()) {
-                if(!reportedInstances.byUuid.contains(knownInstance.getKey())
-                        && !reportedInstances.byExternalId.contains(knownInstance.getValue())) {
-                    if (checkOnly) {
-                        return true;
-                    } else {
-                        // If this known instance was not reported, schedule a potential restart.
-                        restart(knownInstance.getKey());
-                    }
+    /*
+     * If checkOnly is true, will raise a ContainersOutOfSync exception, indicating this should be reran with checkOnly set to false.
+     */
+    void syncContainers(Map<String, Instance> knownInstances, ReportedInstances reportedInstances, long agentAccountId, long agentId, long hostId,
```

Comment:
Just curious - do we apply any lock when this operation is called? I see that the changes is refactoring of existing code, so my questions applies to original implementation. In this PR I see that we lock on per individual instance basis when we operate on it in ContainerEventCreate.java, but not when we determine if the objects are out of sync